File: evaluate_reverse_polish_notation.py
class Solution:
    # @param tokens, a list of string
    # @return an integer
    def evalRPN(self, tokens):
        if not tokens:
            return None
        stk = []
        ops = ['+', '-', '*', '/']
        while tokens:
            a = tokens.pop(0)
            if a not in ops:
                stk.append(int(a))
            else:
                r = stk.pop()
                l = stk.pop()
                back = self.evaluate(l, r, a)
                stk.append(back)
        return int(stk[0])
    
    def evaluate(self, x, y, op):
        if op == '+':
            return x + y
        elif op == '-':
            return x - y
        elif op == '*':
            return x * y
        else:
            if x * y > 0:
                return x / y
            else:
                return -(-x / y)
        
        

File: cleanCode50/EvaluateReversePolishNotation.java
import java.util.*;

public class EvaluateReversePolishNotation {
    public int evalRPN(String[] tokens) {
        if (tokens == null || tokens.length == 0) {
            return 0;
        }
        Stack<Integer> stack = new Stack<>();
        Set<String> operators = new HashSet<>(Arrays.asList("+", "-", "*", "/"));
        for (String token : tokens) {
            if (operators.contains(token)) {
                int right = stack.pop();
                int left = stack.pop();
                stack.push(eval(left, right, token));
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        return stack.pop();
    }

    private int eval(int a, int b, String token) {
        if (token.equals("+")) {
            return a + b;
        } else if (token.equals("-")) {
            return a - b;
        } else if (token.equals("*")) {
            return a * b;
        } else {
            return a / b;
        }
    }

    public static void main(String[] args) {
        EvaluateReversePolishNotation test = new EvaluateReversePolishNotation();
        String[] tokens = {"4", "13", "5", "/", "+"};
        System.out.println(test.evalRPN(tokens));
    }
}